vllm/model_executor/layers/quantization/utils/fp8_utils.py

```diff
@@ -98,122 +104,187 @@ def rocm_aiter_gemm_w8a8_blockscale_fake(
-                                                   (0, 0, 0, num_pad),
-                                                   "constant", 0)
-        q_input, x_scale = per_token_group_quant_fp8(input_2d,
-                                                     block_size[1],
-                                                     column_major_scales=True)
-        output = w8a8_blockscale_func(q_input, weight, x_scale, weight_scale,
-                                      block_size, input.dtype)
-        if num_pad > 0:
-            output = output[:-num_pad]
-    else:
-        if use_aiter_and_is_supported:
-            q_input, x_scale = aiter_per1x128_quant(
-                input_2d.contiguous(), quant_dtype=rocm_aiter.dtypes.fp8)
-        else:
-            q_input, x_scale = per_token_group_quant_fp8(
-                input_2d, block_size[1], column_major_scales=False)
-
-        output = w8a8_blockscale_func(q_input, weight, x_scale, weight_scale,
-                                      block_size, input.dtype)
-
-    if bias is not None:
-        output = output + bias
-    return output.to(dtype=input.dtype).view(*output_shape)
-
-
-def apply_w8a8_block_fp8_linear_fake(
-    input: torch.Tensor,
-    weight: torch.Tensor,
-    block_size: list[int],
-    weight_scale: torch.Tensor,
-    input_scale: Optional[torch.Tensor] = None,
-    bias: Optional[torch.Tensor] = None,
-    cutlass_block_fp8_supported: bool = CUTLASS_BLOCK_FP8_SUPPORTED,
-    use_aiter_and_is_supported: bool = False,
-) -> torch.Tensor:
-    output_shape = [*input.shape[:-1], weight.shape[0]]
-    return torch.empty(output_shape, dtype=input.dtype, device=input.device)
-
+            self.weight_group_shape,
+            output_dtype=input_2d.dtype)
 
-if not current_platform.is_cpu():
-    direct_register_custom_op(
-        op_name="apply_w8a8_block_fp8_linear",
-        op_func=apply_w8a8_block_fp8_linear,
-        mutates_args=[],
-        fake_impl=apply_w8a8_block_fp8_linear_fake,
-    )
+    def _run_cutlass(
+        self,
+        input_2d: torch.Tensor,
+        weight: torch.Tensor,
+        weight_scale: torch.Tensor,
+    ) -> torch.Tensor:
+        assert self.input_quant_op is not None
+        if self.is_hopper:
+            padded_x = torch.nn.functional.pad(
+                input_2d, (0, 0, 0, -input_2d.shape[0] % 4))
+            q_input, x_scale = self.input_quant_op(padded_x)
+            output = cutlass_scaled_mm(q_input, weight, x_scale, weight_scale,
+                                       list(self.weight_group_shape),
+                                       input_2d.dtype, True)
+            output = output[0:input_2d.shape[0], ...]
+        else:
+            q_input, x_scale = self.input_quant_op(input_2d)
+            output = cutlass_scaled_mm(q_input, weight, x_scale, weight_scale,
+                                       list(self.weight_group_shape),
+                                       input_2d.dtype, False)
```

Comment:
I think we can simplify this a bit (the no-op view will get eliminated by torch.compile)
```suggestion
        if self.is_hopper:
            # We pad unconditionally (even if shape is already divisible by 4)
            # to support dynamic shape for input_2d.shape[0] in torch.compile
            x = torch.nn.functional.pad(
                input_2d, (0, 0, 0, -input_2d.shape[0] % 4))
        else:
            x = input_2d
        
        q_input, x_scale = self.input_quant_op(x)
        output = cutlass_scaled_mm(q_input, weight, x_scale, weight_scale,
                                       list(self.weight_group_shape),
                                       input_2d.dtype, self.is_hopper)
        output = output[0:input_2d.shape[0], ...]
```